### ISSUE TITLE
no-useless-undefined vs. no-null vs. eslint/consistent-return vs. void type in TypeScript

### ISSUE THREAD

**Issue by FunctionDJ** [sindresorhus/eslint-plugin-unicorn#880]
Consider a function that can return a primitive or object, or undefined/null/void:
```typescript
function getThing(): string|undefined {
  if (someCondition) {
    return "hello world";
  }

  return undefined;
}
```

In this case, `no-useless-undefined` kicks in. If i change the return and return type to `null`, `no-null` applies. So i'm expected to use `undefined`, but not explicitly. I'm supposed to just `return` without any value. In TypeScript, this is seen as the special type `void` though. Furthermore, then the `eslint/consistent-return` rule kicks in.

I think the `no-useless-undefined` rule should ignore when being used with a `return` when the return type explicitly can be `undefined` in order to make it explicit.

**Comment by sindresorhus** [comment]
@fisker Could we do this just for TypeScript and not for JavaScript?

**Comment by fisker** [comment]
I'm not sure if AST has that info, I'll check.
